### ISSUE TITLE
client-reader worker can leak FDs and then swallow requests

### ISSUE THREAD

**Issue by ara4n** [matrix-org/synapse#1981]
Split out from https://github.com/vector-im/riot-web/issues/3327.

It may or may not be the same as #1980.

**Comment by ara4n** [comment]
Looks like it was getting stuck on acquiring the linearizer lock for sets of state events, starting at ~8pm last night and then swallowing reqs and leaking FDs until it started to fail all requests, and then I restarted it at 23:45.  This is pretty damning:

![screen shot 2017-03-10 at 11 17 44](https://cloud.githubusercontent.com/assets/1294269/23793451/33a77200-0583-11e7-869f-7e90ea5b5352.png)
<img width="952" alt="screen shot 2017-03-10 at 11 16 38" src="https://cloud.githubusercontent.com/assets/1294269/23793452/33afda1c-0583-11e7-858a-6253e13c5ffd.png">

**Comment by ara4n** [comment]
may be related to #1858

**Comment by erikjohnston** [comment]
From the logs:

```
2017-03-09 20:53:52,658 - twisted - 131 - CRITICAL -  - Unhandled error in Deferred:
2017-03-09 20:53:52,658 - twisted - 131 - CRITICAL -  - 
Traceback (most recent call last):
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1357, in gotResult
    _inlineCallbacks(r, g, deferred)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1347, in _inlineCallbacks
    deferred.errback()
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 500, in errback
    self._startRunCallbacks(fail)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 567, in _startRunCallbacks
    self._runCallbacks()
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 666, in _runCallbacks
    current.result = failure.Failure(captureVars=self.debug)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 206, in __init__
    if isinstance(exc_value, str) and exc_type is None:
RuntimeError: maximum recursion depth exceeded while calling a Python object
```

and its full of:

```
2017-03-09 20:53:48,953 - synapse.util.async - 201 - INFO - POST-14459 - Acquired linearizer lock 'state_resolve_lock' for key frozenset([29319867L, 29319869L])
2017-03-09 20:53:48,953 - synapse.util.async - 208 - INFO - POST-14459 - Releasing linearizer lock 'state_resolve_lock' for key frozenset([29319867L, 29319869L])
2017-03-09 20:53:48,988 - synapse.util.async - 201 - INFO - POST-14459 - Acquired linearizer lock 'state_resolve_lock' for key frozenset([29325139L, 29325140L])
2017-03-09 20:53:48,988 - synapse.util.async - 208 - INFO - POST-14459 - Releasing linearizer lock 'state_resolve_lock' for key frozenset([29325139L, 29325140L])
```

from about 23:47, though they all seem to mostly hit the cache.

**Comment by ara4n** [comment]
@erikjohnston did you mean 20:47?

**Comment by erikjohnston** [comment]
Why, yes, I did, thanks

**Comment by erikjohnston** [comment]
After catching it doing a bunch of CPU work and hitting it with pyflame, it seems to be 73% of its time in `get_order_for_room`

**Comment by erikjohnston** [comment]
35% of that is resolving state groups

**Comment by erikjohnston** [comment]
https://matrix.org/_matrix/media/v1/download/jki.re/LhSjIcCyIwwmSQWcvNEAchZT

**Comment by erikjohnston** [comment]
I have changed it so that `get_order_for_room` is much cheaper, so that should help a lot. https://github.com/matrix-org/synapse/pull/1989

**Comment by ara4n** [comment]
do we know why it got permenantly stuck?

**Comment by richvdh** [comment]
sounds like this got fixed.

**Comment by richvdh** [comment]
(sounds like it *didn't* get fixed (properly), but will be by #2532)
